dashboard/src/app/workspaces/workspace-details/stack-selector/stack-selector.controller.ts

```diff
@@ -0,0 +1,206 @@
+  stackIconLinks: {
+    [stackId: string]: string
+  } = {};
+  stackMachines: {
+    [stackId: string]: Array<{[machineProp: string]: string|number}>
+  } = {};
+
+  /**
+   * Default constructor that is using resource injection
+   * @ngInject for Dependency injection
+   */
+  constructor($filter: ng.IFilterService, $log: ng.ILogService, lodash: _.LoDashStatic, cheStack: CheStack, cheEnvironmentRegistry: CheEnvironmentRegistry) {
+    this.$filter = $filter;
+    this.$log = $log;
+    this.lodash = lodash;
+    this.cheStack = cheStack;
+    this.cheEnvironmentRegistry = cheEnvironmentRegistry;
+
+    this.fetchStacks().finally(() => {
+      this.buildFilteredList();
+    });
+  }
+
+  /**
+   * Get list of available stacks.
+   */
+  fetchStacks(): ng.IPromise<any> {
+    return this.cheStack.fetchStacks().then(() => {
+      this.stacks = angular.copy(this.cheStack.getStacks());
+      this.updateStacks();
+    }, (error: any) => {
+      if (error.status === 304) {
+        this.stacks = angular.copy(this.cheStack.getStacks());
+        this.updateStacks();
+      } else {
+        this.$log.error(error);
+      }
+    });
+  }
+
+  /**
+   * For each stack get machines and for each machine cast memory limit to GB.
+   * Get stack icons.
+   */
+  updateStacks(): void {
+    this.stacks.forEach((stack: IStackSelectorItem) => {
+      // get icon link
+      let findLink = this.lodash.find(stack.links, (link: che.IStackLink) => {
+        return link.rel === 'get icon link';
+      });
+      if (findLink) {
+        this.stackIconLinks[stack.id] = findLink.href;
+      }
+
+      // get machines memory limits
+      let defaultEnv = stack.workspaceConfig.defaultEnv,
+          environment = stack.workspaceConfig.environments[defaultEnv],
+          environmentManager = this.getEnvironmentManager(environment.recipe.type),
+          machines = environmentManager.getMachines(environment);
+      this.stackMachines[stack.id] = [];
+      machines.forEach((machine: any) => {
+        let memoryLimitGB = this.memoryToGB(environmentManager.getMemoryLimit(machine));
+        this.stackMachines[stack.id].push({
+          name: machine.name,
+          memoryLimitGB: memoryLimitGB
+        });
+      });
+
+      // set multi machine field
+      stack.isMultiMachine = machines.length > 1;
+    });
+  }
+
+  /**
+   * Returns environment manager specified by recipe type.
+   *
+   * @param recipeType {string} recipe type
+   * @return {EnvironmentManager}
+   */
+  getEnvironmentManager(recipeType: string): EnvironmentManager {
+    if (!this.environmentManagers[recipeType]) {
+      this.environmentManagers[recipeType] = this.cheEnvironmentRegistry.getEnvironmentManager(recipeType);
+    }
+
+    return this.environmentManagers[recipeType];
+  }
+
+  /**
+   * Cast bytes to GB.
+   *
+   * @param memoryLimitBytes {number} memory limit in bytes
+   * @return {number}
+   */
+  memoryToGB(memoryLimitBytes: number): number {
```

Comment:
We don't have already functions performing this kind of convert ?
maybe it should be a component shared by many other components (or a angular filter)